feature(mobile): hash assets & sync via checksum

fyfrey opened the pull request:
**Server needs to be at version v1.59.0 or higher (to provide checksums of asset), else the app won't work at all**

**Removes duplicates shown in app due to backup via CLI or changing app installation method etc.**

**Download button in gallery viewer now works as expected: A local re-sync is triggered after download completes, if checksums match, assets are merged, UI updates, download button disappears. No duplicates are produced.**

**Details:** 
- Drop deviceId, no longer use deviceAssetId of remote assets
- Match/merge local+remote assets purely by their checksum
- Skips duplicates assets encountered on the device (same DB constraint as on server: unique ownerId+checksum)
- Error handling: Assets that cannot be read/hashed are skipped (thus, not shown in the app) and a warning is logged
- Use `crypto` package on iOS, native SHA1 on Android
- On Android, hashing is performed in separate thread (UI is smooth)
- Assets only need to be hashes once. Logout/login keeps asset hashes
- Clearing app data / uninstall also removes hashes (need to rehash)
- After first app start showing local assets is delayed until ALL local assets are hashed (a tiny spinner on the bottom tab icon indicates that processing is active)
- Assets are hashed in batches; if app is closed (by user or system) during initial hashing, only the currently active batch is lost and hashing continues from there.

fyfrey committed:
compare different sha1 implementations

fyfrey committed:
remove openssl sha1

fyfrey committed:
sync via checksum

fyfrey committed:
hash assets in batches

fyfrey committed:
hash in background, show spinner in tab

fyfrey committed:
undo tmp changes

fyfrey committed:
migrate by clearing assets

fyfrey committed:
ignore duplicate assets

alextran1502 commented:
Here is my first result of the test, upon singing in this is what I see

```
Restarted application in 1,128ms.
flutter: ╔══════════════════════════════════════════════════════╗
flutter: ║                 ISAR CONNECT STARTED                 ║
flutter: ╟──────────────────────────────────────────────────────╢
flutter: ║         Open the link to connect to the Isar         ║
flutter: ║        Inspector while this build is running.        ║
flutter: ╟──────────────────────────────────────────────────────╢
flutter: ║ https://inspect.isar.dev/3.1.0+1/#/54115/Sa4OPtyxgv0 ║
flutter: ╚══════════════════════════════════════════════════════╝
flutter: \^[[90m[🌎 Easy Localization] [DEBUG] Localization initialized<…>
flutter: \^[[90m[🌎 Easy Localization] [DEBUG] Start<…>
flutter: \^[[90m[🌎 Easy Localization] [DEBUG] Init state<…>
flutter: \^[[90m[🌎 Easy Localization] [DEBUG] Build<…>
flutter: \^[[90m[🌎 Easy Localization] [DEBUG] Init Localization Delegate<…>
flutter: \^[[90m[🌎 Easy Localization] [DEBUG] Init provider<…>
flutter: Current themeMode system
flutter: \^[[90m[🌎 Easy Localization] [DEBUG] Load Localization Delegate<…>
flutter: \^[[90m[🌎 Easy Localization] [DEBUG] Load asset from assets/i18n<…>
flutter: App Init Completed
flutter: \^[[90m[🌎 Easy Localization] [DEBUG] Load asset from assets/i18n<…>
flutter: \^[[34m[🌎 Easy Localization] [WARNING] Unable to load asset: "assets/i18n/en.json".
The asset does not exist or has empty data.<…>
flutter: \^[[90m[🌎 Easy Localization] [DEBUG] Load asset from assets/i18n<…>
flutter: Attempting to connect to websocket
flutter: [SEVERE] [2023-05-28 21:17:42.962400] Tried to modify a provider while the widget tree was building.
flutter: If you are encountering this error, chances are you tried to modify a provider
flutter: in a widget life-cycle, such as but not limited to:
flutter: - build
flutter: - initState
flutter: - dispose
flutter: - didUpdateWidget
flutter: - didChangeDepedencies
flutter:
flutter: Modifying a provider inside those life-cycles is not allowed, as it could
flutter: lead to an inconsistent UI state. For example, two widgets could listen to the
flutter: same provider, but incorrectly receive different states.
flutter:
flutter:
flutter: To fix this problem, you have one of two solutions:
flutter: - (preferred) Move the logic for modifying your provider outside of a widget
flutter:   life-cycle. For example, maybe you could update your provider inside a button's
flutter:   onPressed instead.
flutter:
flutter: - Delay your modification, such as by encasuplating the modification
flutter:   in a `Future(() {...})`.
flutter:   This will perform your update after the widget tree is done building.
flutter: [SEVERE] [2023-05-28 21:17:42.965392] Tried to modify a provider while the widget tree was building.
flutter: If you are encountering this error, chances are you tried to modify a provider
flutter: in a widget life-cycle, such as but not limited to:
flutter: - build
flutter: - initState
flutter: - dispose
flutter: - didUpdateWidget
flutter: - didChangeDepedencies
flutter:
flutter: Modifying a provider inside those life-cycles is not allowed, as it could
flutter: lead to an inconsistent UI state. For example, two widgets could listen to the
flutter: same provider, but incorrectly receive different states.
flutter:
flutter:
flutter: To fix this problem, you have one of two solutions:
flutter: - (preferred) Move the logic for modifying your provider outside of a widget
flutter:   life-cycle. For example, maybe you could update your provider inside a button's
flutter:   onPressed instead.
flutter:
flutter: - Delay your modification, such as by encasuplating the modification
flutter:   in a `Future(() {...})`.
flutter:   This will perform your update after the widget tree is done building.
flutter: [SEVERE] [2023-05-28 21:17:42.968119] At least listener of the StateNotifier Instance of 'AssetNotifier' threw an exception
flutter: when the notifier tried to update its state.
flutter:
flutter: The exceptions thrown are:
flutter:
flutter: Tried to modify a provider while the widget tree was building.
flutter: If you are encountering this error, chances are you tried to modify a provider
flutter: in a widget life-cycle, such as but not limited to:
flutter: - build
flutter: - initState
flutter: - dispose
flutter: - didUpdateWidget
flutter: - didChangeDepedencies
flutter:
flutter: Modifying a provider inside those life-cycles is not allowed, as it could
flutter: lead to an inconsistent UI state. For example, two widgets could listen to the
flutter: same provider, but incorrectly receive different states.
flutter:
flutter:
flutter: To fix this problem, you have one of two solutions:
flutter: - (preferred) Move the logic for modifying your provider outside of a widget
flutter:   life-cycle. For example, maybe you could update your provider inside a button's
flutter:   onPressed instead.
flutter:
flutter: - Delay your modification, such as by encasuplating the modification
flutter:   in a `Future(() {...})`.
flutter:   This will perform your update after the widget tree is done building.
flutter: #0      _UncontrolledProviderScopeElement._debugCanModifyProviders (package:flutter_riverpod/src/framework.dart:343:7)
flutter: #1      ProviderElementBase._notifyListeners.<anonymous closure> (package:riverpod/src/framework/element.dart:480:34)
flutter: #2      ProviderElementBase._notifyListeners (package:riverpod/src/framework/element.dart:482:8)
flutter: #3      ProviderElementBase.setState (package:riverpod/src/framework/element.dart:150:7)
flutter: #4      StateNotifier.state= (package:state_notifier/state_notifier.dart:225:31)
flutter: #5      AssetNotifier.getAllAsset (package:immich_mobile/shared/providers/asset.provider.dart:64:7)
flutter: #6      HomePage.build.<anonymous closure> (package:immich_mobile/modules/home/views/home_page.dart:51:43)
flutter: #7      _EffectHookState.scheduleEffect (package:flutter_hooks/src/primitives.dart:229:27)
flutter: #8      _EffectHookState.initHook (package:flutter_hooks/src/primitives.dart:209:5)
flutter: #9      _extension#0._createHookState (package:flutter_hooks/src/framework.dart:318:9)
flutter: #10     _extension#0._appendHook (package:flutter_hooks/src/framework.dart:329:20)
flutter: #11     HookElement._use (package:flutter_hooks/src/framework.dart:438:7)
flutter: #12     Hook.use (package:flutter_hooks/src/framework.dart:139:45)
flutter: #13     use (package:flutter_hooks/src/framework.dart:18:32)
flutter: #14     useEffect (package:flutter_hooks/src/primitives.dart:191:3)
flutter: #15     HomePage.build (package:immich_mobile/modules/home/views/home_page.dart:48:5)
flutter: #16     _ConsumerState.build (package:flutter_riverpod/src/consumer.dart:477:19)
flutter: #17     StatefulElement.build (package:flutter/src/widgets/framework.dart:5198:27)
flutter: #18     ConsumerStatefulElement.build (package:flutter_riverpod/src/consumer.dart:540:20)
flutter: #19     HookElement.build (package:flutter_hooks/src/framework.dart:416:27)
flutter: #20     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5086:15)
flutter: #21     StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:5251:11)
flutter: #22     Element.rebuild (package:flutter/src/widgets/framework.dart:4805:7)
flutter: #23     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:5068:5)
flutter: #24     StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:5242:11)
flutter: #25     ComponentElement.mount (package:flutter/src/widgets/framework.dart:5062:5)
flutter: #26     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #27     Element.updateChild (package:flutter/src/widgets/framework.dart:3708:18)
flutter: #28     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5111:16)
flutter: #29     Element.rebuild (package:flutter/src/widgets/framework.dart:4805:7)
flutter: #30     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:5068:5)
flutter: #31     ComponentElement.mount (package:flutter/src/widgets/framework.dart:5062:5)
flutter: #32     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #33     Element.updateChild (package:flutter/src/widgets/framework.dart:3708:18)
flutter: #34     SingleChildRenderObjectElement.mount (package:flutter/src/widgets/framework.dart:6435:14)
flutter: #35     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #36     Element.updateChild (package:flutter/src/widgets/framework.dart:3708:18)
flutter: #37     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5111:16)
flutter: #38     Element.rebuild (package:flutter/src/widgets/framework.dart:4805:7)
flutter: #39     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:5068:5)
flutter: #40     ComponentElement.mount (package:flutter/src/widgets/framework.dart:5062:5)
flutter: #41     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #42     Element.updateChild (package:flutter/src/widgets/framework.dart:3708:18)
flutter: #43     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5111:16)
flutter: #44     Element.rebuild (package:flutter/src/widgets/framework.dart:4805:7)
flutter: #45     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:5068:5)
flutter: #46     ComponentElement.mount (package:flutter/src/widgets/framework.dart:5062:5)
flutter: #47     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #48     MultiChildRenderObjectElement.inflateWidget (package:flutter/src/widgets/framework.dart:6570:36)
flutter: #49     MultiChildRenderObjectElement.mount (package:flutter/src/widgets/framework.dart:6582:32)
flutter: #50     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #51     Element.updateChild (package:flutter/src/widgets/framework.dart:3708:18)
flutter: #52     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5111:16)
flutter: #53     Element.rebuild (package:flutter/src/widgets/framework.dart:4805:7)
flutter: #54     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:5068:5)
flutter: #55     ComponentElement.mount (package:flutter/src/widgets/framework.dart:5062:5)
flutter: #56     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #57     Element.updateChild (package:flutter/src/widgets/framework.dart:3708:18)
flutter: #58     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5111:16)
flutter: #59     StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:5251:11)
flutter: #60     Element.rebuild (package:flutter/src/widgets/framework.dart:4805:7)
flutter: #61     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:5068:5)
flutter: #62     StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:5242:11)
flutter: #63     ComponentElement.mount (package:flutter/src/widgets/framework.dart:5062:5)
flutter: #64     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #65     Element.updateChild (package:flutter/src/widgets/framework.dart:3708:18)
flutter: #66     SingleChildRenderObjectElement.mount (package:flutter/src/widgets/framework.dart:6435:14)
flutter: #67     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #68     Element.updateChild (package:flutter/src/widgets/framework.dart:3708:18)
flutter: #69     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5111:16)
flutter: #70     Element.rebuild (package:flutter/src/widgets/framework.dart:4805:7)
flutter: #71     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:5068:5)
flutter: #72     ComponentElement.mount (package:flutter/src/widgets/framework.dart:5062:5)
flutter: #73     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #74     Element.updateChild (package:flutter/src/widgets/framework.dart:3708:18)
flutter: #75     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5111:16)
flutter: #76     Element.rebuild (package:flutter/src/widgets/framework.dart:4805:7)
flutter: #77     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:5068:5)
flutter: #78     ComponentElement.mount (package:flutter/src/widgets/framework.dart:5062:5)
flutter: #79     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #80     Element.updateChild (package:flutter/src/widgets/framework.dart:3708:18)
flutter: #81     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5111:16)
flutter: #82     Element.rebuild (package:flutter/src/widgets/framework.dart:4805:7)
flutter: #83     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:5068:5)
flutter: #84     ComponentElement.mount (package:flutter/src/widgets/framework.dart:5062:5)
flutter: #85     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #86     Element.updateChild (package:flutter/src/widgets/framework.dart:3708:18)
flutter: #87     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5111:16)
flutter: #88     Element.rebuild (package:flutter/src/widgets/framework.dart:4805:7)
flutter: #89     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:5068:5)
flutter: #90     ComponentElement.mount (package:flutter/src/widgets/framework.dart:5062:5)
flutter: #91     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #92     Element.updateChild (package:flutter/src/widgets/framework.dart:3708:18)
flutter: #93     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5111:16)
flutter: #94     Element.rebuild (package:flutter/src/widgets/framework.dart:4805:7)
flutter: #95     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:5068:5)
flutter: #96     ComponentElement.mount (package:flutter/src/widgets/framework.dart:5062:5)
flutter: #97     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #98     MultiChildRenderObjectElement.inflateWidget (package:flutter/src/widgets/framework.dart:6570:36)
flutter: #99     MultiChildRenderObjectElement.mount (package:flutter/src/widgets/framework.dart:6582:32)
flutter: #100    Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #101    Element.updateChild (package:flutter/src/widgets/framework.dart:3708:18)
flutter: #102    ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5111:16)
flutter: #103    Element.rebuild (package:flutter/src/widgets/framework.dart:4805:7)
flutter: #104    ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:5068:5)
flutter: #105    ComponentElement.mount (package:flutter/src/widgets/framework.dart:5062:5)
flutter: #106    Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #107    Element.updateChild (package:flutter/src/widgets/framework.dart:3708:18)
flutter: #108    ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5111:16)
flutter: #109    StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:5251:11)
flutter: #110    Element.rebuild (package:flutter/src/widgets/framework.dart:4805:7)
flutter: #111    ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:5068:5)
flutter: #112    StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:5242:11)
flutter: #113    ComponentElement.mount (package:flutter/src/widgets/framework.dart:5062:5)
flutter: #114    Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #115    Element.updateChild (package:flutter/src/widgets/framework.dart:3708:18)
flutter: #116    ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5111:16)
flutter: #117    StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:5251:11)
flutter: #118    Element.rebuild (package:flutter/src/widgets/framework.dart:4805:7)
flutter: #119    ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:5068:5)
flutter: #120    StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:5242:11)
flutter: #121    ComponentElement.mount (package:flutter/src/widgets/framework.dart:5062:5)
flutter: #122    Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #123    Element.updateChild (package:flutter/src/widgets/framework.dart:3708:18)
flutter: #124    ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5111:16)
flutter: #125    Element.rebuild (package:flutter/src/widgets/framework.dart:4805:7)
flutter: #126    ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:5068:5)
flutter: #127    ComponentElement.mount (package:flutter/src/widgets/framework.dart:5062:5)
flutter: #128    Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #129    Element.updateChild (package:flutter/src/widgets/framework.dart:3708:18)
flutter: #130    ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5111:16)
flutter: #131    StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:5251:11)
flutter: #132    Element.rebuild (package:flutter/src/widgets/framework.dart:4805:7)
flutter: #133    ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:5068:5)
flutter: #134    StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:5242:11)
flutter: #135    ComponentElement.mount (package:flutter/src/widgets/framework.dart:5062:5)
flutter: #136    Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #137    Element.updateChild (package:flutter/src/widgets/framework.dart:3708:18)
flutter: #138    SingleChildRenderObjectElement.mount (package:flutter/src/widgets/framework.dart:6435:14)
flutter: #139    Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #140    Element.updateChild (package:flutter/src/widgets/framework.dart:3708:18)
flutter: #141    ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5111:16)
flutter: #142    Element.rebuild (package:flutter/src/widgets/framework.dart:4805:7)
flutter: #143    ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:5068:5)
flutter: #144    ComponentElement.mount (package:flutter/src/widgets/framework.dart:5062:5)
flutter: #145    Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #146    Element.updateChild (package:flutter/src/widgets/framework.dart:3708:18)
flutter: #147    SingleChildRenderObjectElement.mount (package:flutter/src/widgets/framework.dart:6435:14)
flutter: #148    Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #149    Element.updateChild (package:flutter/src/widgets/framework.dart:3708:18)
flutter: #150    ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5111:16)
flutter: #151    StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:5251:11)
flutter: #152    Element.rebuild (package:flutter/src/widgets/framework.dart:4805:7)
flutter: #153    ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:5068:5)
flutter: #154    StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:5242:11)
flutter: #155    ComponentElement.mount (package:flutter/src/widgets/framework.dart:5062:5)
flutter: #156    Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #157    Element.updateChild (package:flutter/src/widgets/framework.dart:3708:18)
flutter: #158    ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5111:16)
flutter: #159    StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:5251:11)
flutter: #160    Element.rebuild (package:flutter/src/widgets/framework.dart:4805:7)
flutter: #161    ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:5068:5)
flutter: #162    StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:5242:11)
flutter: #163    ComponentElement.mount (package:flutter/src/widgets/framework.dart:5062:5)
flutter: #164    Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #165    Element.updateChild (package:flutter/src/widgets/framework.dart:3708:18)
flutter: #166    ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5111:16)
flutter: #167    Element.rebuild (package:flutter/src/widgets/framework.dart:4805:7)
flutter: #168    ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:5068:5)
flutter: #169    ComponentElement.mount (package:flutter/src/widgets/framework.dart:5062:5)
flutter: #170    Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #171    Element.updateChild (package:flutter/src/widgets/framework.dart:3708:18)
flutter: #172    ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5111:16)
flutter: #173    Element.rebuild (package:flutter/src/widgets/framework.dart:4805:7)
flutter: #174    ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:5068:5)
flutter: #175    ComponentElement.mount (package:flutter/src/widgets/framework.dart:5062:5)
flutter: #176    Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #177    Element.updateChild (package:flutter/src/widgets/framework.dart:3708:18)
flutter: #178    ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5111:16)
flutter: #179    Element.rebuild (package:flutter/src/widgets/framework.dart:4805:7)
flutter: #180    ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:5068:5)
flutter: #181    ComponentElement.mount (package:flutter/src/widgets/framework.dart:5062:5)
flutter: #182    Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #183    Element.updateChild (package:flutter/src/widgets/framework.dart:3708:18)
flutter: #184    ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5111:16)
flutter: #185    StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:5251:11)
flutter: #186    Element.rebuild (package:flutter/src/widgets/framework.dart:4805:7)
flutter: #187    ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:5068:5)
flutter: #188    StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:5242:11)
flutter: #189    ComponentElement.mount (package:flutter/src/widgets/framework.dart:5062:5)
flutter: #190    Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #191    Element.updateChild (package:flutter/src/widgets/framework.dart:3708:18)
flutter: #192    ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5111:16)
flutter: #193    Element.rebuild (package:flutter/src/widgets/framework.dart:4805:7)
flutter: #194    ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:5068:5)
flutter: #195    ComponentElement.mount (package:flutter/src/widgets/framework.dart:5062:5)
flutter: #196    Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #197    Element.updateChild (package:flutter/src/widgets/framework.dart:3708:18)
flutter: #198    ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5111:16)
flutter: #199    StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:5251:11)
flutter: #200    Element.rebuild (package:flutter/src/widgets/framework.dart:4805:7)
flutter: #201    ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:5068:5)
flutter: #202    StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:5242:11)
flutter: #203    ComponentElement.mount (package:flutter/src/widgets/framework.dart:5062:5)
flutter: #204    Element.inflateWidget (package:flutter/src/widgets/framework.dart:3971:16)
flutter: #205    Element.updateChild (package:flutter/src/widgets/framework.dart:3708:18)
flutter: #206    _LayoutBuilderElement._layout.layoutCallback (package:flutter/src/widgets/layout_builder.dart:135:18)
flutter: #207    BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2720:19)
flutter: #208    _LayoutBuilderElement._layout (package:flutter/src/widgets/layout_builder.dart:153:12)
flutter: #209    RenderObject.invokeLayoutCallback.<anonymous closure> (package:flutter/src/rendering/object.dart:2506:59)
flutter: #210    PipelineOwner._enableMutationsToDirtySubtrees (package:flutter/src/rendering/object.dart:1062:15)
flutter: #211    RenderObject.invokeLayoutCallback (package:flutter/src/rendering/object.dart:2506:14)
flutter: #212    RenderConstrainedLayoutBuilder.rebuildIfNecessary (package:flutter/src/widgets/layout_builder.dart:228:7)
flutter: #213    _RenderLayoutBuilder.performLayout (package:flutter/src/widgets/layout_builder.dart:313:5)
flutter: #214    RenderObject.layout (package:flutter/src/rendering/object.dart:2395:7)
flutter: #215    RenderBox.layout (package:flutter/src/rendering/box.dart:2386:11)
flutter: #216    RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:122:14)
flutter: #217    RenderObject.layout (package:flutter/src/rendering/object.dart:2395:7)
flutter: #218    RenderBox.layout (package:flutter/src/rendering/box.dart:2386:11)
flutter: #219    RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:122:14)
flutter: #220    RenderObject.layout (package:flutter/src/rendering/object.dart:2395:7)
flutter: #221    RenderBox.layout (package:flutter/src/rendering/box.dart:2386:11)
flutter: #222    RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:122:14)
flutter: #223    RenderObject.layout (package:flutter/src/rendering/object.dart:2395:7)

flutter: #224    RenderBox.layout (package:flutter/src/rendering/box.dart:2386:11)
flutter: #225    RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:122:14)
flutter: #226    RenderObject.layout (package:flutter/src/rendering/object.dart:2395:7)
flutter: #227    RenderBox.layout (package:flutter/src/rendering/box.dart:2386:11)
flutter: #228    RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:122:14)
flutter: #229    RenderObject.layout (package:flutter/src/rendering/object.dart:2395:7)
flutter: #230    RenderBox.layout (package:flutter/src/rendering/box.dart:2386:11)
flutter: #231    RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:122:14)
flutter: #232    RenderObject.layout (package:flutter/src/rendering/object.dart:2395:7)
flutter: #233    RenderBox.layout (package:flutter/src/rendering/box.dart:2386:11)
flutter: #234    RenderOffstage.performLayout (package:flutter/src/rendering/proxy_box.dart:3749:14)
flutter: #235    RenderObject.layout (package:flutter/src/rendering/object.dart:2395:7)
flutter: #236    RenderBox.layout (package:flutter/src/rendering/box.dart:2386:11)
flutter: #237    RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:122:14)
flutter: #238    RenderObject.layout (package:flutter/src/rendering/object.dart:2395:7)
flutter: #239    RenderBox.layout (package:flutter/src/rendering/box.dart:2386:11)
flutter: #240    _RenderTheaterMixin.performLayout (package:flutter/src/widgets/overlay.dart:832:15)
flutter: #241    RenderObject._layoutWithoutResize (package:flutter/src/rendering/object.dart:2234:7)
flutter: #242    PipelineOwner.flushLayout (package:flutter/src/rendering/object.dart:1016:18)
flutter: #243    RendererBinding.drawFrame (package:flutter/src/rendering/binding.dart:492:19)
flutter: #244    WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:905:13)
flutter: #245    RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:358:5)
flutter: #246    SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1284:15)
flutter: #247    SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1214:9)
flutter: #248    SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:1072:5)
flutter: #249    _invoke (dart:ui/hooks.dart:142:13)
flutter: #250    PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:359:5)
flutter: #251    _drawFrame (dart:ui/hooks.dart:112:31)
flutter:
flutter:
flutter:
flutter: Established Websocket Connection
flutter: [INFO] [2023-05-28 21:17:43.334814] Found 27 device albums
flutter: [INFO] [2023-05-28 21:17:43.340185] Syncing a new local album to DB: Immich Upload
flutter: Files: 0, Total time: 19ms, total bytes: 0, 0.0 mb/s
flutter: [INFO] [2023-05-28 21:17:43.403893] 0 assets already existed in DB, to upsert 154
flutter: [INFO] [2023-05-28 21:17:43.412499] Upserted 154 assets into the DB
flutter: [INFO] [2023-05-28 21:17:43.568266] Added a new local album to DB: Immich Upload
flutter: [INFO] [2023-05-28 21:17:43.568308] Syncing a new local album to DB: Raw dng
flutter: Files: 0, Total time: 2ms, total bytes: 0, 0.0 mb/s
flutter: [INFO] [2023-05-28 21:17:43.670639] 0 assets already existed in DB, to upsert 1
flutter: [INFO] [2023-05-28 21:17:43.672476] Upserted 1 assets into the DB
flutter: [INFO] [2023-05-28 21:17:43.675925] Added a new local album to DB: Raw dng
flutter: [INFO] [2023-05-28 21:17:43.675955] Syncing a new local album to DB: RAW
flutter: Files: 0, Total time: 6ms, total bytes: 0, 0.0 mb/s
flutter: [INFO] [2023-05-28 21:17:43.823018] 1 assets already existed in DB, to upsert 13
flutter: [INFO] [2023-05-28 21:17:43.825504] Upserted 13 assets into the DB
flutter: [INFO] [2023-05-28 21:17:43.828129] Added a new local album to DB: RAW
flutter: [INFO] [2023-05-28 21:17:43.828157] Syncing a new local album to DB: Filmborn
flutter: Files: 0, Total time: 1ms, total bytes: 0, 0.0 mb/s
flutter: [INFO] [2023-05-28 21:17:43.842970] 0 assets already existed in DB, to upsert 7
flutter: [INFO] [2023-05-28 21:17:43.843917] Upserted 7 assets into the DB
flutter: [INFO] [2023-05-28 21:17:43.846109] Added a new local album to DB: Filmborn
flutter: [INFO] [2023-05-28 21:17:43.846130] Syncing a new local album to DB: Recents
flutter: Hashing 128 files / 737302431 b took 8801ms: 79.89392022140566 mb/s
```

It only shows a subset of the assets and these logs keep popping up

```
flutter: Hashing 128 files / 737302431 b took 8801ms: 79.89392022140566 mb/s
flutter: Hashing 128 files / 1079047653 b took 12677ms: 81.17535953652627 mb/s
flutter: Hashing 128 files / 255192621 b took 3462ms: 70.29770317275974 mb/s
flutter: Hashing 128 files / 870399162 b took 10582ms: 78.44238573246294 mb/s
flutter: Hashing 128 files / 877237691 b took 10581ms: 79.06616154335339 mb/s
flutter: Hashing 128 files / 633660883 b took 7822ms: 77.25723720639296 mb/s
flutter: Hashing 128 files / 861836489 b took 10255ms: 80.14737440283156 mb/s
flutter: Hashing 128 files / 469221561 b took 5800ms: 77.15250886719802 mb/s
```

Then I go into the backup page to make sure that I choose the album/folder that containers all the assets on the device. It then loaded some of the remote assets but still not showing all the local assets. And then some errors showed in the log.

```
flutter: Stop listening to event on_upload_success
flutter: [INFO] [2023-05-28 21:19:36.663317] Found 33 local albums
flutter: _getBackupAlbumsInfo takes 1487ms
flutter: Start listening to event on_upload_success
flutter: [SEVERE] [2023-05-28 21:20:01.897210] Exception: File error asset 4B7E6E2F-958D-4E0D-AE67-4659903F0C9D/L0/001
flutter: [INFO] [2023-05-28 21:20:01.926086] Upserted 10 assets into the DB
flutter: refreshRemoteAssets full took 42ms
flutter: [INFO] [2023-05-28 21:20:02.117785] Found 27 device albums
flutter: [INFO] [2023-05-28 21:20:02.145134] Syncing a new local album to DB: Recents
error = Error Domain=PHPhotosErrorDomain Code=-1 "(null)"
flutter: [SEVERE] [2023-05-28 21:20:04.693966] Exception: File error asset 4B7E6E2F-958D-4E0D-AE67-4659903F0C9D/L0/001
```

Maybe the `File error asset 4B7E6E2F-958D-4E0D-AE67-4659903F0C9D/L0/001` causing an issue with hashing the entire library on the phone?

fyfrey committed:
error handling

fyfrey committed:
trigger sync/merge after download and update view

jrasm91 commented on mobile/lib/shared/services/hash.service.dart:
```diff
@@ -0,0 +1,183 @@
+    for (int i = 0; i < assets.length; i++) {
+      if (hashes[i] != null) continue;
+      final file = await assets[i].originFile;
+      if (file == null) {
```
What if the last file in this list is `null`? Does the last queue of `toHash` never get processed?

jrasm91 commented on mobile/lib/shared/services/hash.service.dart:
```diff
@@ -0,0 +1,183 @@
+
+  Future<List<Asset>> hashAssets(List<AssetEntity> assets) async {
+    final Stopwatch totalTime = Stopwatch()..start();
+    const int batchFileCount = 128;
+    const int batchDataSize = 1024 * 1024 * 1024; // 1GB
```
I've always seen constants in uppercase, does the code everywhere use camel case for un-changeable constants?

fyfrey replied:
dart uses camelCase also for constants.. not sure why but otherwise the style checker complains

jrasm91 replied:
Cool, sounds good.

fyfrey commented:
@alextran1502 please test again :-)

jrasm91 commented on mobile/lib/shared/services/hash.service.dart:
```diff
@@ -0,0 +1,183 @@
+      hashes[i] = da;
+      if (toHash.length == batchFileCount ||
+          bytes >= batchDataSize ||
+          (i + 1 == assets.length && toHash.isNotEmpty)) {
```
This line seems buggy. There are several situations where the for loop would `continue` above on potentially the last item, and this code would never execute. It needs to be run here and after the for loop exits.

fyfrey replied:
good catch!

jrasm91 commented on mobile/lib/shared/services/hash.service.dart:
```diff
@@ -0,0 +1,183 @@
+          "Hashing ${toAdd.length} files / $bytes b took ${sw.elapsedMilliseconds}ms: ${(bytes / (1024 * 1024.0)) / sw.elapsedMilliseconds * 1000} mb/s",
+        );
+        final validHashes =
+            anyNull ? toAdd.where((e) => e.hash.isNotEmpty).toList() : toAdd;
```
What is the purpose in tracking `anyNull`. It seems like a performance optimization to avoid filtering the list here. The code would probably be clearer if you deleted lines 102, 108, and updated this to just always filter out empty values.

jrasm91 reviewed (commented):
The hashing service is quite hard to follow IMO. From what I can tell it looks right for the most part. It would be easier to read if it was broken into separate methods (there are quite a few ternary statements, for loops, and nested for loops), didn't use abbreviated variable names, avoided the single line if + return statements, etc. I think synchronization in the mobile app is very challenging. It makes sense that it will be complex and bugs usually live in the most complicated parts of the system. Anything you can do to improve readability and reduce the complexity is worthwhile IMO, especially here.

Thanks for taking on this implementation. I think it is coming along nicely. It is definitely more complicated than I thought it would be. I think it makes sense to keep a cache table of local checksums. The server does have an API endpoint to check assets by checksums, but it looks like you don't have to use that, since you can do a local database compare instead. Is that correct?

fyfrey committed:
review feedback improvements

fyfrey commented on mobile/lib/shared/services/hash.service.dart:
```diff
@@ -0,0 +1,183 @@
+  }
+
+  Future<List<Asset>> hashAssets(List<AssetEntity> assets) async {
+    final Stopwatch totalTime = Stopwatch()..start();
```
immediate execute the start() method (that returns void) but still allow to assign the constructor's value to the variable

fyfrey commented on mobile/lib/modules/asset_viewer/providers/image_viewer_page_state.provider.dart:
```diff
@@ -34,6 +39,7 @@ class ImageViewerStateNotifier extends StateNotifier<ImageViewerPageState> {
         toastType: ToastType.success,
         gravity: ToastGravity.BOTTOM,
       );
+      _albumService.refreshDeviceAlbums();
```
nope, it's needed for the refresh when clicking the download button in the gallery

fyfrey commented on mobile/android/app/src/main/kotlin/com/example/mobile/BackgroundServicePlugin.kt:
```diff
@@ -70,9 +76,57 @@ class BackgroundServicePlugin : FlutterPlugin, MethodChannel.MethodCallHandler {
+                            digest.update(buf, 0, len)
+                            hashes[i] = digest.digest()
+                        } catch (e: Exception) {
+                            // skip this file
```
added logging (only accessible by developers, though)

fyfrey commented on mobile/android/app/src/main/kotlin/com/example/mobile/BackgroundServicePlugin.kt:
```diff
@@ -70,9 +76,57 @@ class BackgroundServicePlugin : FlutterPlugin, MethodChannel.MethodCallHandler {
+                }
+            }
+            "digestFile" -> {
+                val args = call.arguments<ArrayList<*>>()!!
```
removed the single file case as it's not needed

jrasm91 reviewed (approved):
Looks much better!

alextran1502 commented:
Hi Fynn, below is the result of the second round of testing.

I can see that the file hash is all calculated and present in the Isar database. 6200 assets took about 10 minutes to run through them all.

Then I encountered this errors while trying to upsert those assets into the Isar database.

```bash
flutter: [INFO] [2023-05-31 13:29:59.300951] 83 assets already existed in DB, to upsert 319
flutter: [WARNING] [2023-05-31 13:29:59.302321] Failed to upsert 319 assets into the DB: IsarError: Unique index violated.
flutter: [SEVERE] [2023-05-31 13:29:59.302859] Failed to add new local album Selfies to DB: IsarError: Object "{
flutter:   "remoteId": "N/A",
flutter:   "localId": "E3828AF6-6D1E-4554-976E-2C0B592D16CE/L0/001",
flutter:   "checksum": "v15H1+Izrs9xu2/ZxT0Qk97UHRQ=",
flutter:   "ownerId": "-5252019566449266991",
flutter:   "livePhotoVideoId": "N/A",
flutter:   "fileCreatedAt": "2020-10-05 15:00:15.000",
flutter:   "fileModifiedAt": "2022-11-27 10:08:25.000",
flutter:   "updatedAt": "2022-11-27 10:08:25.000",
flutter:   "durationInSeconds": 0,
flutter:   "type": "AssetType.image",
flutter:   "fileName": "",
flutter:   "isFavorite": false,
flutter:   "isRemote: false,
flutter:   "storage": AssetState.local,
flutter:   "width": 2320,
flutter:   "height": 3088,
flutter:   "isArchived": false
flutter: }" has no id and can therefore not be linked. Make sure to .put() objects before you use them in links.
flutter: [INFO] [2023-05-31 13:29:59.325192] Syncing a new local album to DB: Imaging Edge Mobile
flutter: [INFO] [2023-05-31 13:29:59.337170] 0 assets already existed in DB, to upsert 13
flutter: [WARNING] [2023-05-31 13:29:59.337513] Failed to upsert 13 assets into the DB: IsarError: Unique index violated.
flutter: [SEVERE] [2023-05-31 13:29:59.337806] Failed to add new local album Imaging Edge Mobile to DB: IsarError: Object "{
flutter:   "remoteId": "N/A",
flutter:   "localId": "022D824E-A5D7-44B1-AF27-F90C6A7CE4D1/L0/001",
flutter:   "checksum": "/MR97SNQh8wzqUh/4dbFv5ctdCg=",
flutter:   "ownerId": "-5252019566449266991",
flutter:   "livePhotoVideoId": "N/A",
flutter:   "fileCreatedAt": "2022-12-15 17:43:21.000",
flutter:   "fileModifiedAt": "2022-12-15 21:16:42.000",
flutter:   "updatedAt": "2022-12-15 21:16:42.000",
flutter:   "durationInSeconds": 0,
flutter:   "type": "AssetType.image",
flutter:   "fileName": "",
flutter:   "isFavorite": false,
flutter:   "isRemote: false,
flutter:   "storage": AssetState.local,
flutter:   "width": 1616,
flutter:   "height": 1080,
flutter:   "isArchived": false
flutter: }" has no id and can therefore not be linked. Make sure to .put() objects before you use them in links.
```

Let me know if you need more information

fyfrey commented:
> Hi Fynn, below is the result of the second round of testing.
> 
> I can see that the file hash is all calculated and present in the Isar database. 6200 assets took about 10 minutes to run through them all.
> 
> Then I encountered this errors while trying to upsert those assets into the Isar database.
> 
> ```bash
> flutter: [INFO] [2023-05-31 13:29:59.300951] 83 assets already existed in DB, to upsert 319
> flutter: [WARNING] [2023-05-31 13:29:59.302321] Failed to upsert 319 assets into the DB: IsarError: Unique index violated.
> flutter: [SEVERE] [2023-05-31 13:29:59.302859] Failed to add new local album Selfies to DB: IsarError: Object "{
> flutter:   "remoteId": "N/A",
> flutter:   "localId": "E3828AF6-6D1E-4554-976E-2C0B592D16CE/L0/001",
> flutter:   "checksum": "v15H1+Izrs9xu2/ZxT0Qk97UHRQ=",
> flutter:   "ownerId": "-5252019566449266991",
> flutter:   "livePhotoVideoId": "N/A",
> flutter:   "fileCreatedAt": "2020-10-05 15:00:15.000",
> flutter:   "fileModifiedAt": "2022-11-27 10:08:25.000",
> flutter:   "updatedAt": "2022-11-27 10:08:25.000",
> flutter:   "durationInSeconds": 0,
> flutter:   "type": "AssetType.image",
> flutter:   "fileName": "",
> flutter:   "isFavorite": false,
> flutter:   "isRemote: false,
> flutter:   "storage": AssetState.local,
> flutter:   "width": 2320,
> flutter:   "height": 3088,
> flutter:   "isArchived": false
> flutter: }" has no id and can therefore not be linked. Make sure to .put() objects before you use them in links.
> flutter: [INFO] [2023-05-31 13:29:59.325192] Syncing a new local album to DB: Imaging Edge Mobile
> flutter: [INFO] [2023-05-31 13:29:59.337170] 0 assets already existed in DB, to upsert 13
> flutter: [WARNING] [2023-05-31 13:29:59.337513] Failed to upsert 13 assets into the DB: IsarError: Unique index violated.
> flutter: [SEVERE] [2023-05-31 13:29:59.337806] Failed to add new local album Imaging Edge Mobile to DB: IsarError: Object "{
> flutter:   "remoteId": "N/A",
> flutter:   "localId": "022D824E-A5D7-44B1-AF27-F90C6A7CE4D1/L0/001",
> flutter:   "checksum": "/MR97SNQh8wzqUh/4dbFv5ctdCg=",
> flutter:   "ownerId": "-5252019566449266991",
> flutter:   "livePhotoVideoId": "N/A",
> flutter:   "fileCreatedAt": "2022-12-15 17:43:21.000",
> flutter:   "fileModifiedAt": "2022-12-15 21:16:42.000",
> flutter:   "updatedAt": "2022-12-15 21:16:42.000",
> flutter:   "durationInSeconds": 0,
> flutter:   "type": "AssetType.image",
> flutter:   "fileName": "",
> flutter:   "isFavorite": false,
> flutter:   "isRemote: false,
> flutter:   "storage": AssetState.local,
> flutter:   "width": 1616,
> flutter:   "height": 1080,
> flutter:   "isArchived": false
> flutter: }" has no id and can therefore not be linked. Make sure to .put() objects before you use them in links.
> ```
> 
> Let me know if you need more information

Thanks for testing again. Never encountered that issue - it's really annoying that I don't have access to a Mac to test on iOS^^
I'll look at the code to check how this can happen and might come back to you if I need more info.

zoodyy committed:
hash in background isolate on iOS

zoodyy committed:
rework linking assets with existing from DB

fyfrey committed:
fine-grained errors on unique index violation

fyfrey committed:
hash lenth validation

fyfrey commented:
@alextran1502 I've reworked a small, but central part of the sync service where assets are linked with any existing data from the DB. further, I've added more fine-grained logging for your case of a unique index violation.
Please re-test on your device (in release mode so that hashing is somewhat quick). If this still has errors, you can also try a debug build afterwards, I've added a number of `assert`s to check the correct operation. Thanks!

zoodyy committed:
revert compute in background on iOS

fyfrey commented:
> Hello @fyfrey, another round of testing. There is an error once it starts hashing.
> 
> ```shell
> flutter: [SEVERE] [2023-06-06 08:27:35.224289] Invalid argument(s): Illegal argument in isolate message: object is unsendable - Library:'package:isar/src/native/isar_impl.dart' Class: IsarImpl (see restrictions listed at `SendPort.send()` documentation for more information)
> flutter:  <- Instance of 'HashService' (from package:immich_mobile/shared/services/hash.service.dart)
> flutter:  <- Context num_variables: 1
> flutter:  <- Closure: (List<String>) => Future<List<Uint8List?>> (from dart:core)
> flutter:  <- Context num_variables: 2
> flutter:  <- Closure: () => FutureOr<List<Uint8List?>> (from dart:core)
> flutter:  <- Instance of '_RemoteRunner<List<Uint8List?>>' (from dart:isolate)
> flutter:
> ```
> 
> It doesn't provide the line where the error occurs, so I am a bit loss on how to troubleshoot this :-?

seems something cannot be send to another isolate to compute the slow dart hash in background. I've reverted that change. Please test again.. sorry for the inconvenience!

fyfrey committed:
ignore duplicate assets on device

alextran1502 commented:
Hi @fyfrey, good news. It is now working :)

Here are a few observations that I want to confirm are the expected behavior

1. Upon logging in to the app, it will load the remote address first, then once the backup albums are selected, it will start the hashing process. Assets that are hashed will then be shown on the timeline.
2. Duplicated assets in the same album will be skipped.
3. Files that cannot be read will be skipped.
4. I don't see hashing in the Recents album, an asset is hashed one by one, so I got a few thousand log messages. Is this intended?

fyfrey committed:
fix bug with batching based on accumulated size

alextran1502 commented on mobile/lib/modules/home/views/home_page.dart:
```diff
@@ -47,11 +47,11 @@ class HomePage extends HookConsumerWidget {
-        ref.watch(sharedAlbumProvider.notifier).getAllSharedAlbums();
-        ref.watch(serverInfoProvider.notifier).getServerVersion();
+        ref.read(websocketProvider.notifier).connect();
+        Future(() => ref.read(assetProvider.notifier).getAllAsset());
```
Why `Future` is used here?

fyfrey replied:
getAllAssets is called during build but can modify state causing a rebuild, flutter throws an error/warning and suggest to put the call in a future so it's delayed after painting
